File: bot.py
#!/usr/bin/env python3
import asyncio, time, hmac, hashlib, json
from datetime import datetime, timezone
import httpx
from pathlib import Path

CURRENT_FILE = Path("current_listing.json")
if not CURRENT_FILE.exists():
    raise FileNotFoundError(f"{CURRENT_FILE} not found")

with open(CURRENT_FILE, "r", encoding="utf-8") as f:
    listing = json.load(f)
print(f"[BOT] Loaded listing: {listing}")

API_KEY          = listing["api_key"]
API_SECRET       = listing["api_secret"]
SYMBOL           = listing["symbol"].upper()
QUOTE_AMOUNT     = float(listing["quote_amount"])
LISTING_TIME     = listing["listing_time"]
PRICE_MARKUP_PCT = float(listing.get("price_markup_pct", 20))
PROFIT_PCT       = float(listing.get("profit_pct", 200))

REST_URL = "https://api.mexc.com"

def sign(params: dict, secret: str) -> str:
    qs = "&".join(f"{k}={params[k]}" for k in params)
    return hmac.new(secret.encode(), qs.encode(), hashlib.sha256).hexdigest()

def log_attempts(attempts):
    print("\n📊 Tabela prób:\n")
    hdr = f"{'Nr':<3} | {'Wysłano':<23} | {'Odpowiedź':<23} | {'Lat(ms)':<8} | {'Status':<5} | {'Qty':<8} | {'Msg'}"
    print(hdr)
    print("-" * len(hdr))
    for i,a in enumerate(attempts,1):
        print(f"{i:<3} | {a['sent']:<23} | {a['recv']:<23} | {a['lat']:>7.2f} | {a['status']:<5} | {a['exec_qty']:<8.6f} | {a['msg']}")

async def get_server_offset(client):
    r = await client.get(f"{REST_URL}/api/v3/time")
    return r.json()["serverTime"] - int(time.time()*1000)

async def warmup(client):
    await client.get(f"{REST_URL}/api/v3/time")
    ts = int(time.time()*1000)-100_000
    dummy = {
        "symbol": SYMBOL,
        "side": "BUY",
        "type": "MARKET",
        "quoteOrderQty": "0.000001",
        "recvWindow": "2000",
        "timestamp": str(ts)
    }
    dummy["signature"] = sign(dummy, API_SECRET)
    await client.post(f"{REST_URL}/api/v3/order", params=dummy, headers={"X-MEXC-APIKEY":API_KEY})

async def prepare_buy(client):
    r = await client.get(f"{REST_URL}/api/v3/depth", params={"symbol":SYMBOL,"limit":5})
    asks = r.json().get("asks",[])
    if not asks:
        return None
    price = float(asks[0][0])
    limit = round(price*(1+PRICE_MARKUP_PCT/100),8)
    print(f"[PREP] market={price} → limit={limit}")
    return {
        "limit_price":limit,
        "template_base": {
            "symbol": SYMBOL,
            "side": "BUY",
            "type": "LIMIT",
            "price": str(limit),
            "quantity": None,
            "timeInForce": "IOC",
            "recvWindow": "5000"
        },
        "url": f"{REST_URL}/api/v3/order",
        "headers": {"X-MEXC-APIKEY":API_KEY}
    }

def busy_wait_until(target_ms:int):
    while int(time.time()*1000) < target_ms:
        time.sleep(0.0005)

async def fetch_executed_qty(client, order_id):
    # Pobiera rzeczywisty executedQty z giełdy (GET /order)
    params = {
        "symbol": SYMBOL,
        "orderId": order_id,
        "timestamp": str(int(time.time()*1000))
    }
    params["signature"] = sign(params, API_SECRET)
    resp = await client.get(f"{REST_URL}/api/v3/order", params=params, headers={"X-MEXC-APIKEY":API_KEY})
    d = resp.json()
    qty = float(d.get("executedQty", "0"))
    return qty, d

async def place_buy(client, build, offset, send_at, qty, success_event):
    busy_wait_until(send_at)
    if success_event.is_set():
        return None
    p = build["template_base"].copy()
    p["quantity"] = str(qty)
    p["timestamp"] = str(int(time.time()*1000)+offset)
    p = {k: p[k] for k in ["symbol","side","type","price","quantity","timeInForce","recvWindow","timestamp"]}
    p["signature"] = sign(p, API_SECRET)
    sent = datetime.now().strftime("%H:%M:%S.%f")[:-3]
    start = time.perf_counter()
    resp = await client.post(build["url"], params=p, headers=build["headers"])
    lat = (time.perf_counter()-start)*1000
    d = resp.json()
    order_id = d.get("orderId")
    qty_exec = 0.0
    msg = d.get("msg","")
    status = "ERR"
    # Jeśli jest orderId — pobierz executedQty z GET /order
    if order_id:
        try:
            await asyncio.sleep(0.5)  # daj giełdzie czas na fill (opcjonalnie: powtarzaj kilka razy)
            qty_exec, order_details = await fetch_executed_qty(client, order_id)
            status = "OK" if qty_exec > 0 else "NOFILL"
            msg += " | "+order_details.get("status", "")
        except Exception as e:
            msg += f" | ERR GET /order: {e}"
    result = {"sent":sent,"recv":datetime.now().strftime("%H:%M:%S.%f")[:-3],"lat":lat,"status":status,"exec_qty":qty_exec,"msg":msg}
    if status == "OK" and qty_exec > 0:
        success_event.set()
    return result

async def place_market(client, offset, send_at, amount, success_event):
    busy_wait_until(send_at)
    if success_event.is_set():
        return None
    ts = str(int(time.time()*1000)+offset)
    params = {
        "symbol": SYMBOL,
        "side": "BUY",
        "type": "MARKET",
        "quoteOrderQty": str(amount),
        "recvWindow": "5000",
        "timestamp": ts
    }
    params["signature"] = sign(params, API_SECRET)
    sent = datetime.now().strftime("%H:%M:%S.%f")[:-3]
    start = time.perf_counter()
    resp = await client.post(f"{REST_URL}/api/v3/order", params=params, headers={"X-MEXC-APIKEY":API_KEY})
    lat = (time.perf_counter()-start)*1000
    d = resp.json()
    order_id = d.get("orderId")
    qty_exec = 0.0
    msg = d.get("msg","")
    status = "ERR"
    if order_id:
        try:
            await asyncio.sleep(0.5)
            qty_exec, order_details = await fetch_executed_qty(client, order_id)
            status = "OK" if qty_exec > 0 else "NOFILL"
            msg += " | "+order_details.get("status", "")
        except Exception as e:
            msg += f" | ERR GET /order: {e}"
    result = {"sent":sent,"recv":datetime.now().strftime("%H:%M:%S.%f")[:-3],"lat":lat,"status":status,"exec_qty":qty_exec,"msg":msg}
    if status == "OK" and qty_exec > 0:
        success_event.set()
    return result

async def main():
    print(f"[INFO] {SYMBOL} @ {LISTING_TIME}, amount={QUOTE_AMOUNT}, markup={PRICE_MARKUP_PCT}%, profit={PROFIT_PCT}%")
    async with httpx.AsyncClient(http2=True) as client:
        offset=await get_server_offset(client)
        print(f"[SYNC] offset: {offset}ms")

        t0=int(datetime.fromisoformat(LISTING_TIME).astimezone(timezone.utc).timestamp()*1000)
        now=lambda:int(time.time()*1000)
        while now()<t0-4000: await asyncio.sleep(0.01)
        offset=await get_server_offset(client)
        print(f"[SYNC] offset (przed zakupem): {offset}ms")

        while now()<t0-5000: await asyncio.sleep(0.01)
        build = await prepare_buy(client)
        await warmup(client)
        t0_local = t0
        buy_times = [t0_local-10, t0_local-5, t0_local]
        success_event = asyncio.Event()
        atts = []
        rem = QUOTE_AMOUNT
        bought = 0

        if build is not None:
            qty = round(rem / build["limit_price"], 6)
            tasks = [
                place_buy(client, build, offset, buy_times[i], qty, success_event)
                for i in range(3)
            ]
        else:
            print("[BOT] Próba market (orderbook pusty/lub fail limitów)...")
            tasks = [
                place_market(client, offset, buy_times[i], rem, success_event)
                for i in range(3)
            ]

        results = await asyncio.gather(*tasks)
        results = [r for r in results if r is not None]
        log_attempts(results)

        for r in results:
            if r["status"]=="OK" and r["exec_qty"]>0:
                if build is not None:
                    bought = r["exec_qty"]*build["limit_price"]
                else:
                    bought = r["exec_qty"]
                break

        if bought<=0:
            print("[BOT] no buy")
            return

        # SELL
        if build is not None:
            sell_price=round((build["limit_price"])*(1+PROFIT_PCT/100),8)
            sell_qty=bought/(build["limit_price"])
        else:
            sell_price = 0
            sell_qty = bought
        print(f"[BOT] SELL {sell_qty}@{sell_price}")
        sp={
            "symbol": SYMBOL,
            "side": "SELL",
            "type": "LIMIT",
            "price": str(sell_price),
            "quantity": str(round(sell_qty,6)),
            "timeInForce":"GTC",
            "recvWindow":"5000",
            "timestamp":str(int(time.time()*1000)+offset)
        }
        sp = {k: sp[k] for k in ["symbol","side","type","price","quantity","timeInForce","recvWindow","timestamp"]}
        sp["signature"]=sign(sp,API_SECRET)
        sent=datetime.now().strftime("%H:%M:%S.%f")[:-3]
        start=time.perf_counter()
        resp=await client.post(f"{REST_URL}/api/v3/order",params=sp,headers={"X-MEXC-APIKEY":API_KEY})
        lat=(time.perf_counter()-start)*1000
        d=resp.json(); status="OK" if "orderId" in d else "ERR"
        log_attempts([{"sent":sent,"recv":datetime.now().strftime("%H:%M:%S.%f")[:-3],"lat":lat,"status":status,"exec_qty":sell_qty,"msg":d.get("msg","")}])

if __name__=="__main__":
    asyncio.run(main())

File: my_new_dir/bot.go
package main

import (
    "bytes"
    "crypto/hmac"
    "crypto/sha256"
    "encoding/json"
    "fmt"
    "io/ioutil"
    "log"
    "net/http"
    "os"
    "sort"
    "strconv"
    "time"
)

const REST_URL = "https://api.mexc.com"

type Listing struct {
    ID             string  `json:"id"`
    APIKey         string  `json:"api_key"`
    APISecret      string  `json:"api_secret"`
    Symbol         string  `json:"symbol"`
    QuoteAmount    float64 `json:"quote_amount"`
    ListingTime    string  `json:"listing_time"`
    PriceMarkupPct float64 `json:"price_markup_pct"`
    ProfitPct      float64 `json:"profit_pct"`
}

func must(err error) {
    if err != nil { log.Fatal(err) }
}

func sign(params map[string]string, secret string) string {
    keys := make([]string,0,len(params))
    for k:=range params { keys = append(keys,k) }
    sort.Strings(keys)
    var buf bytes.Buffer
    for i,k := range keys {
        if i>0 { buf.WriteString("&") }
        buf.WriteString(fmt.Sprintf("%s=%s", k, params[k]))
    }
    mac := hmac.New(sha256.New, []byte(secret))
    mac.Write(buf.Bytes())
    return fmt.Sprintf("%x", mac.Sum(nil))
}

func httpReq(method, url string, headers, qs map[string]string) []byte {
    req, _ := http.NewRequest(method, url, nil)
    q := req.URL.Query()
    for k,v := range qs { q.Set(k,v) }
    req.URL.RawQuery = q.Encode()
    for k,v := range headers { req.Header.Set(k,v) }
    resp, err := http.DefaultClient.Do(req)
    must(err)
    defer resp.Body.Close()
    b, _ := ioutil.ReadAll(resp.Body)
    return b
}

func main() {
    if len(os.Args)<2 {
        log.Fatal("brak listing_id")
    }
    lid := os.Args[1]
    // 1) załaduj current_listing.json
    b,_ := ioutil.ReadFile("current_listing.json")
    var l Listing
    must(json.Unmarshal(b,&l))
    log.Printf("[INFO] %s @ %s, amount=%.4f, markup=%.2f%%, profit=%.2f%%",
        l.Symbol, l.ListingTime, l.QuoteAmount, l.PriceMarkupPct, l.ProfitPct)

    // 2) oblicz T0
    t0, _ := time.Parse(time.RFC3339, l.ListingTime)
    t0ms := t0.UTC().UnixNano()/1e6

    // 3) sync offset
    resp := httpReq("GET", REST_URL+"/api/v3/time", nil, nil)
    var tm struct{ ServerTime int64 `json:"serverTime"` }
    must(json.Unmarshal(resp,&tm))
    offset := tm.ServerTime - time.Now().UnixNano()/1e6
    log.Printf("[SYNC] offset=%dms", offset)

    // 4) dummy
    warmTs := strconv.FormatInt(time.Now().UnixNano()/1e6-offset-100000,10)
    dummy := map[string]string{
        "symbol":l.Symbol,"side":"BUY","type":"MARKET",
        "quoteOrderQty":"0.000001","recvWindow":"2000","timestamp":warmTs,
    }
    dummy["signature"]=sign(dummy,l.APISecret)
    httpReq("POST", REST_URL+"/api/v3/order", map[string]string{"X-MEXC-APIKEY":l.APIKey}, dummy)

    // 5) pobierz book
    dep := httpReq("GET", REST_URL+"/api/v3/depth", nil, map[string]string{"symbol":l.Symbol,"limit":"5"})
    var D struct{ Asks [][]string `json:"asks"` }
    must(json.Unmarshal(dep,&D))
    price,_ := strconv.ParseFloat(D.Asks[0][0],64)
    limitPrice := math.Round(price*(1+l.PriceMarkupPct/100)*1e8)/1e8
    log.Printf("[PREP] market=%.8f → limit=%.8f", price, limitPrice)

    // 6) qty
    qty := math.Round(l.QuoteAmount/limitPrice*1e6)/1e6

    // 7) próby
    offsets := []int64{-10,-5,0}
    var done bool
    type R struct{ Sent string; Lat float64; Status string; Qty float64 }
    var results []R

    for _,off := range offsets {
        target := t0ms + off
        for time.Now().UnixNano()/1e6 < target-offset { /*busy*/ }
        if done { break }
        params := map[string]string{
            "symbol":l.Symbol,"side":"BUY","type":"LIMIT",
            "price":fmt.Sprintf("%.8f",limitPrice),
            "quantity":fmt.Sprintf("%.6f",qty),
            "timeInForce":"IOC","recvWindow":"5000",
            "timestamp":strconv.FormatInt(time.Now().UnixNano()/1e6+offset,10),
        }
        params["signature"]=sign(params,l.APISecret)
        start:=time.Now()
        body:=httpReq("POST",REST_URL+"/api/v3/order",
            map[string]string{"X-MEXC-APIKEY":l.APIKey}, params)
        lat := float64(time.Since(start).Microseconds())/1000
        var d map[string]interface{}
        json.Unmarshal(body,&d)
        status,filled := "ERR",0.0
        if _, ok := d["orderId"]; ok {
            filled = d["executedQty"].(float64)
            if filled>0{ status="OK"; done=true }
        }
        results = append(results,R{Sent:start.Format("15:04:05.000"), Lat:lat, Status:status, Qty:filled})
    }

    // 8) log
    for i,r := range results {
        log.Printf("%d: %s | %.2fms | %s | qty=%.6f",i+1,r.Sent,r.Lat,r.Status,r.Qty)
    }
    if !done { log.Println("[BOT] no buy"); return }

    // 9) SELL
    sellPrice:=math.Round(limitPrice*(1+l.ProfitPct/100)*1e8)/1e8
    sellQty:=qty
    log.Printf("[BOT] SELL %f @ %.8f",sellQty,sellPrice)
    sp:=map[string]string{
        "symbol":l.Symbol,"side":"SELL","type":"LIMIT",
        "price":fmt.Sprintf("%.8f",sellPrice),
        "quantity":fmt.Sprintf("%.6f",sellQty),
        "timeInForce":"GTC","recvWindow":"5000",
        "timestamp":strconv.FormatInt(time.Now().UnixNano()/1e6+offset,10),
    }
    sp["signature"]=sign(sp,l.APISecret)
    httpReq("POST",REST_URL+"/api/v3/order",map[string]string{"X-MEXC-APIKEY":l.APIKey},sp)
    log.Println("[SELL] done")
}
